### PR TITLE
feat: Add `NullSentinels` mixin for Optional-field null handling

### LINKED ISSUE #8: Add configurable value-remapping mixin
## Summary

The current `Metric` class has a `_empty_field_to_none` model validator that automatically converts empty string fields to `None` during parsing. This behavior is enabled by default for consistency with fgpyo's `Metric`, but could be improved with a more flexible, opt-in approach.

## Context

https://github.com/fulcrumgenomics/fgmetric/pull/6#discussion_r2686278842

> It might be nice to have `""` be a `ClassVar` so that users could override it with their own special "empty token". 

The current implementation hardcodes the empty string to `None` conversion, which doesn't allow for:
1. Disabling the behavior when not needed
2. Customizing which values map to `None`
3. Supporting multiple input representations (e.g., both "" and "None" -> `None`)
4. Handling other sentinel values like "NA" -> `np.nan`

## Suggested Solution

Create a mixin class (e.g., `RemapValues`) that:

1. Uses a dict class variable for arbitrary value mappings instead of a single "empty token"
2. Generalizes to support cases like:
    - multiple string representations of null: {"": None, "None": None} 
    - mixed null/NaN handling: {"": None, "NA": np.nan} 
3. (Maybe?) Is opt-in rather than enabled by default on the base Metric class

### PR DESCRIPTION
**Pull request by msto**
Closes #8.

## Summary

Clears the way for #10 (a base class with no default converters mixed in, so users can opt in/out as necessary for performance/functionality).

This PR moves the "empty string becomes None" validation off `Metric` and into its own mixin, much like the delimited list handling. It also renames the `collections` module to `converters` for consistency.

## Summary by CodeRabbit

* **New Features**
  * Added configurable “null sentinel” handling to `Metric`, converting configured empty-string (and other sentinel) inputs to `None` for optional fields during validation.
  * Sentinel substitution respects field aliases/validation aliases and distinguishes optional fields from required ones.
  * Improved interaction with delimited list parsing (sentinel-to-`None` happens before list splitting).

* **Tests**
  * Expanded coverage for default behavior, custom sentinel sets, alias scenarios, inheritance overrides, and edge cases.

* **Chores**
  * Updated converter utility exports and removed related re-exports from collections.

### CODE REVIEW

File: fgmetric/metric.py [hunk 1]
```diff
@@ -32,15 +32,18 @@ class Metric(
     header of the file. Subclasses should define their fields using Pydantic field annotations.
 
     `Metric` includes the following custom serialization/deserialization behaviors:
-    1. **Empty fields as None.** Any empty field in a file will be represented as `None` on the
-       deserialized model.
+    1. **Null sentinels.** Input strings in the `null_sentinels` class variable are treated as
+       `None` on Optional fields before field validation. Defaults to `frozenset({""})`, so any
+       empty Optional field is represented as `None` on the deserialized model.
```
Comment (msto) [inline fgmetric/metric.py]:
```suggestion
    1. **Null sentinels.** Empty strings in Optional fields are converted to `None` before field validation. The sentinel values converted to `None` can be overridden by the `null_sentinels` class variable. 
```

File: fgmetric/converters/_null_sentinels.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+                continue
+            keys.add(name)
+            keys |= _validation_keys(info)
+        cls._optional_field_keys = keys
+
+    @final
+    @model_validator(mode="before")
+    @classmethod
+    def _substitute_null_sentinels(cls, data: Any) -> Any:
+        """Substitute incoming sentinel strings on Optional fields with `None`."""
```
Comment (ameynert) [inline fgmetric/converters/_null_sentinels.py:107]: [suggestion](https://pullpo.io/cc?l=suggestion) 
According to Claude: 

  The deleted `_empty_field_to_none` carried a valuable NB comment explaining the `list[T]` vs `list[T] | None` empty-string behavior. Since that interaction is subtle and now spans two files (mixin + `DelimitedList`), consider porting a condensed version of that note onto the mixin so the next reader doesn't have to re-derive it.

```
  # NB: "Before" validators (mode="before") run before field validators such as
  # `DelimitedList._split_lists()`. Empty strings in Optional fields will always be converted to
  # `None` before any field validators.
  # For example, for delimited list parsing:
  #   - When a field is defined as `list[T] | None`, this converts "" → None before _split_lists
  #     sees it.
  #   - When a field is defined as `list[T]`, "" passes through unchanged, then _split_lists
  #     converts "" → [].
```

File: tests/test_null_sentinels.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    class Model(NullSentinels):
+        null_sentinels: ClassVar[frozenset[str]] = frozenset({""})
+
+        name: str
+        value: int | None
+
+    result = Model.model_validate({"name": "foo", "value": ""})
+    assert result.value is None
+
```
Comment (ameynert) [inline tests/test_null_sentinels.py:35]: [suggestion](https://pullpo.io/cc?l=suggestion) 
Add a test on a `Metric` subclass that has both a `list[T]` and a `list[T] | None` field to show how the behaviour differs between the two.

```python
def test_optional_list_field_empties_to_none_but_required_list_empties_to_empty_list() -> None:
    """
    The null-sentinel substitution and `DelimitedList` splitting compose correctly on `""`.

   `_substitute_null_sentinels` (model `mode="before"`) runs before `_split_lists` (field
    validator), so an empty string in an Optional list field becomes `None`, while an empty
    string in a required list field is left untouched and `_split_lists` turns it into `[]`.
    """

    class FakeMetric(Metric):
        required: list[int]
        optional: list[int] | None
  
    result = FakeMetric.model_validate({"required": "", "optional": ""})
    assert result.required == []      # not Optional -> untouched -> split to []
    assert result.optional is None    # Optional -> "" substituted to None before splitting

    # Non-empty values still parse as delimited lists in both cases.
    populated = FakeMetric.model_validate({"required": "1,2", "optional": "3,4"})
    assert populated.required == [1, 2]
    assert populated.optional == [3, 4]
```

File: fgmetric/converters/_null_sentinels.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+    The substitution is scoped to Optional fields. Non-Optional fields are not touched, which
+    keeps this mixin composable with field-specific handling (e.g., `DelimitedList`'s treatment
+    of `list[T]` fields).
+
+    Class Variables:
+        null_sentinels: The set of input strings that should be treated as null on Optional
+            fields. Defaults to an empty set (no substitution).
+
+    Examples:
```
Comment (ameynert) [inline fgmetric/converters/_null_sentinels.py:52]: [suggestion](https://pullpo.io/cc?l=suggestion) 
Both examples subclass `Metric` and the first sets `null_sentinels = frozenset({""})`, which is already `Metric`'s default, making the override redundant. Since this docstring lives on the *mixin*, an example using `NullSentinels` directly (or dropping the redundant override) would more clearly illustrate what the mixin itself contributes.

File: fgmetric/metric.py [hunk 5]
```diff
@@ -32,15 +32,18 @@ class Metric(
...
+       `null_sentinels` class variable.
     2. **Delimited lists.** Any field typed as `list[T]` will be parsed from and serialized to a
        delimited string. The list delimiter may be controlled by the `collection_delimiter` class
        variable.
 
     Class Variables:
         collection_delimiter: A single-character delimiter used to split and join `list` fields
             during serialization/deserialization.
+        null_sentinels: The set of input strings that should be treated as `None` on Optional
+            fields during validation. Defaults to `frozenset({""})`.
```
Comment (ameynert) [inline fgmetric/metric.py:46]: [suggestion](https://pullpo.io/cc?l=suggestion) 
Add a "see `NullSentinels`" pointer here.

**Review by ameynert — APPROVED** [review]
